File: benchmarks/c/binary_trees.c
// cc -O2 binary_trees.c -o binary_trees_c
#include <stdio.h>
#include <stdint.h>

int64_t tree_sum(int64_t depth, int64_t value) {
    if (depth == 0) return value;
    return tree_sum(depth - 1, value * 2) + tree_sum(depth - 1, value * 2 + 1);
}

int main(void) {
    int64_t s = tree_sum(18, 1);
    printf("binary_trees(depth=18) = %lld\n", (long long)s);
    return 0;
}

File: benchmarks/go/binary_trees.go
// Benchmark: binary trees — pure recursion; mirrors c/binary_trees.c exactly.
// Build: go build -o bin/binary_trees_go go/binary_trees.go
package main

import "fmt"

func treeSum(depth, value int64) int64 {
	if depth == 0 {
		return value
	}
	return treeSum(depth-1, value*2) + treeSum(depth-1, value*2+1)
}

func main() {
	s := treeSum(18, 1)
	fmt.Printf("binary_trees(depth=21) = %d\n", s)
}
